### PR TITLE
Refactor useSoftkey

### PR DESCRIPTION
**Pull request by stephanebisson**
* Make useSoftkey more declarative
* Allow popup to have their own softkeys while
  they are open and have the parent's component
  softkey continue working once the popups are
  closed.

### CODE REVIEW

File: src/hooks/useSoftkey.js [hunk 1]
```diff
@@ -1,4 +1,21 @@
...
+    return () => { newContext && softkey.dispatch({ type: 'pop' }) }
+  }, [newContext])
+
+  useEffect(() => {
+    if (config) {
+      softkey.dispatch({ type: 'set', config })
+    }
+  }, dependencies)
+  return softkey
+}
```
Comment (hueitan) [inline src/hooks/useSoftkey.js]: route to a new page, popup a new component, always `newContext=true`

in which case do you think the `newContext` needed to be `false` here?

Comment (stephanebisson) [reply src/hooks/useSoftkey.js]: In this commit, it is `false` in `<ArticleSection>` to signify to it contributes to the context created by `<Article>` and doesn't replace it. 

Another way to achieve something similar would be to use a string like the `origin` param to specify which view a component belongs to. All components of the article view could specify `article` to share a context.

Comment (stephanebisson) [reply src/hooks/useSoftkey.js]: The second option proposed here has the advantage of being independent from the call order.

Comment (hueitan) [reply src/hooks/useSoftkey.js]: 👍 I see the latest commit, `origin` looks clearer to define which context it is.

File: src/hooks/useArticlePagination.js [hunk 2]
```diff
@@ -1,22 +1,20 @@
...
 ) => {
   const [currentSection, setCurrentSection] = useState(findSection(article.toc, subTitle))
   const [isLastPage, setIsLastPage] = useState(0)
   const prop = 'scrollLeft'
   const numOfSection = article.sections.length
 
-  useKeys({
-    ArrowDown: () => {
-      if (isTocShown) return
+  useSoftkey('Article', {
```
Comment (hueitan) [inline src/hooks/useArticlePagination.js]: 👍👍👍 This is really awesome!

File: src/components/Article.js [hunk 3]
```diff
@@ -14,30 +14,34 @@ const ArticleBody = memo(({ content }) => {
...
-      type: 'setCenterText',
-      value: hasActions ? i18n.i18n('centerkey-select') : ''
-    })
-    if (hasActions) {
-      setNavigation(0)
-    }
-  }, [hasActions])
-
+  const [,, getCurrent] = useNavigation('Article', actionsRef, 'x')
+  const [isVisible, setVisible] = useState(false)
```
Comment (hueitan) [inline src/components/Article.js]: should this rename to which component is visible? or if this is the temporary state, then make the comment so we know it when we reach this again.

File: src/components/ArticleToc.js [hunk 4]
```diff
@@ -6,19 +6,8 @@ import { ListView } from 'components'
 export const ArticleToc = ({ items, close }) => {
```
Comment (hueitan) [inline src/components/ArticleToc.js]: I would also like to rename the close function to be `closeFn`, same as the audio popup, I will do it in my later bug fix pr.

File: src/components/Search.js [hunk 5]
```diff
@@ -5,35 +5,28 @@ import { useNavigation, useSearch, useLanguage, useI18n, useSoftkey } from 'hook
...
-
   const onKeyCenter = () => {
     const { index, key } = getCurrent()
     if (index) {
       window.location.hash = `/article/${lang}/${key}`
     }
   }
 
+  useSoftkey('Search', {
+    left: 'Settings',
```
Comment (hueitan) [inline src/components/Search.js]: `i18n.i18n( 'settings' )`

File: src/components/Article.js (hunk 3 above)
Comment (stephanebisson) [reply src/components/Article.js]: Done

**Pull request merged**